code/go/0chain.net/smartcontract/storagesc/allocation_eventdb.go

```diff
@@ -0,0 +1,239 @@
+			WritePrice:              b.Terms.WritePrice,
+			MinLockDemand:           b.Terms.MinLockDemand,
+			MaxOfferDuration:        b.Terms.MaxOfferDuration,
+			ChallengeCompletionTime: b.Terms.ChallengeCompletionTime,
+		})
+	}
+
+	termsByte, err := json.Marshal(allocationTerms)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling terms: %v", err)
+	}
+
+	alloc := &event.Allocation{
+		AllocationID:               sa.ID,
+		TransactionID:              sa.Tx,
+		DataShards:                 sa.DataShards,
+		ParityShards:               sa.ParityShards,
+		Size:                       sa.Size,
+		Expiration:                 int64(sa.Expiration),
+		Terms:                      string(termsByte),
+		Owner:                      sa.Owner,
+		OwnerPublicKey:             sa.OwnerPublicKey,
+		IsImmutable:                sa.IsImmutable,
+		ReadPriceMin:               sa.ReadPriceRange.Min,
+		ReadPriceMax:               sa.ReadPriceRange.Max,
+		WritePriceMin:              sa.WritePriceRange.Min,
+		WritePriceMax:              sa.WritePriceRange.Max,
+		MaxChallengeCompletionTime: int64(sa.MaxChallengeCompletionTime),
+		ChallengeCompletionTime:    int64(sa.ChallengeCompletionTime),
+		StartTime:                  int64(sa.StartTime),
+		Finalized:                  sa.Finalized,
+		Cancelled:                  sa.Canceled,
+		UsedSize:                   sa.UsedSize,
+		MovedToChallenge:           sa.MovedToChallenge,
+		MovedBack:                  sa.MovedBack,
+		MovedToValidators:          sa.MovedToValidators,
+		Curators:                   strings.Join(sa.Curators, ","),
+		TimeUnit:                   int64(sa.TimeUnit),
+	}
+
+	if sa.Stats != nil {
+		alloc.NumWrites = sa.Stats.NumWrites
+		alloc.NumReads = sa.Stats.NumReads
+		alloc.TotalChallenges = sa.Stats.TotalChallenges
+		alloc.OpenChallenges = sa.Stats.OpenChallenges
+		alloc.SuccessfulChallenges = sa.Stats.SuccessChallenges
+		alloc.FailedChallenges = sa.Stats.FailedChallenges
+		alloc.LatestClosedChallengeTxn = sa.Stats.LastestClosedChallengeTxn
+	}
+
+	return alloc, nil
+}
+
+func emitAddOrOverwriteAllocation(sa *StorageAllocation, balances cstate.StateContextI) error {
+
+	alloc, err := storageAllocationToAllocationTable(sa)
+	if err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(alloc)
+	if err != nil {
+		return fmt.Errorf("error marshalling allocation: %v", err)
+	}
+
+	balances.EmitEvent(event.TypeStats, event.TagAddOrOverwriteAllocation, alloc.AllocationID, string(data))
+
+	return nil
+}
+
+func getStorageAllocationFromDb(id string, balances cstate.StateContextI) (*StorageAllocation, error) {
+
+	alloc, err := balances.GetEventDB().GetAllocation(id)
+	if err != nil {
+		return nil, err
+	}
+
+	sa, err := allocationTableToStorageAllocation(alloc, balances)
+	if err != nil {
+		return nil, err
+	}
+
+	return sa, nil
+}
+
+func getClientAllocationsFromDb(clientID string, balances cstate.StateContextI) ([]*StorageAllocation, error) {
```

Comment:
Pass an `*event.EventDb` object instead of cstate.StateContextI object.

Reply:
done